pystac/item_collection.py

```diff
@@ -0,0 +1,125 @@
+from copy import deepcopy
+from pystac.errors import STACTypeError
+from typing import Any, Dict, Iterator, List, Optional, Sized, Iterable
+
+import pystac
+from pystac.utils import make_absolute_href, is_absolute_href
+from pystac.serialization.identify import identify_stac_object_type
+
+
+class ItemCollection(Sized, Iterable[pystac.Item]):
+    """Implementation of a GeoJSON FeatureCollection whose features are all STAC
+    Items.
+
+    All :class:`~pystac.Item` instances passed to the :class:`~ItemCollection` instance
+    during instantiation are cloned and have their ``"root"`` URL cleared. Instances of
+    this class are iterable and sized (see examples below).
+
+    Any additional top-level fields in the FeatureCollection are retained in
+    :attr:`~ItemCollection.extra_fields` by the :meth:`~ItemCollection.from_dict` and
+    :meth:`~ItemCollection.from_file` methods and will be present in the serialized file
+    from :meth:`~ItemCollection.save_object`.
+
+    Examples:
+
+        Loop over all items in the ItemCollection
+
+        >>> item_collection: ItemCollection = ...
+        >>> for item in item_collection:
+        ...     ...
+
+        Get the number of Items in the ItemCollection
+
+        >>> length: int = len(item_collection)
+
+    """
+
+    items: List[pystac.Item]
+    """The list of :class:`pystac.Item` instances contained in this
+    ``ItemCollection``."""
+
+    extra_fields: Dict[str, Any]
+    """Dictionary containing additional top-level fields for the GeoJSON
+    FeatureCollection."""
+
+    def __init__(
+        self, items: List[pystac.Item], extra_fields: Optional[Dict[str, Any]] = None
+    ):
+        self.items = [item.clone() for item in items]
+        for item in self.items:
+            item.clear_links("root")
```

Comment:
I've been thinking about this. On one hand keeping root and other hierarchical links, is nice because it essentially provides information on where the Items originally came from.  You can see the original root endpoint it came from.

On the other hand, these new Items no longer have a root themselves, nor parents, so what do their inclusion mean?  If you split out this ItemCollection into separate Item files, is that when  you would strip out the hierarchical links?   Or when you edit the file?   If the ItemCollection is simply an unaltered copy from an API then keeping the links probably makes sense.  Not sure the best approach here.

- `root` would be a way of identifying the original catalog the Item came from
- `parent` doesn't seem useful in an ItemCollection at all. I think `root` is better.  The `parent` is either going to be a collection (if from an API), or a sub-catalog if this ItemCollection was created manually from a bunch of Items.
- `collection` links should be kept, as the ItemCollection doesn't contain collections itself, a way is needed to fetch the original collection this Item belonged to
- `

Reply:
Since the `ItemCollection` is such a generic vessel for Items, it seems like the best course here might be to not take any action on the Items at all (other than possible cloning, per the comment above). If/when we introduce other convenience methods for converting to/from `Collection` instance, we could handle the links more specifically. It would also be more transparent to users if we just put Items into the ItemCollection unaltered and let them handle any cleanup they want to do.

Reply:
Agree, since we don't really know what should be done here we should just leave it as is. An ItemCollection is just a collection of Items as they are.

Reply:
Removed the link manipulation in 9b1fed25c879511e91ead274df5d5244520f34b8